### DISCUSSION TITLE
Rename Column Dynamically

### DISCUSSION THREAD

**Discussion by MagDev175** [serenity-is/Serenity#6483] [Help] [answered]
I have a grid that is part of a dialog (like[ this example](https://volkanceylan.gitbooks.io/serenity-guide/content/tutorials/movies/listing_movies_in_person_dialog.html)). I want to change a column name in the grid based on a value in the dialog. So if dialog fieldA is "Y", grid Field1 has a display name of "Y" and if dialog fieldA is "Z", grid Field1 has a display name of "Z". Can someone help with this?

**Comment by stixoffire** [comment]
If you only want to change the display name in the grid , but on the value change 
Subscribed to the the Value Field something like this .. (cant remember off the top of my head) 
// Subscription in you xyzgrid 
```
MyForm.field.change.subcribe((fieldvalue)=> { if (fieldvalue == Z ){this.changeTheColumnName(Z, "what I want")} }); 

protected changeTheColumnName(fieldName: string, newName: string){
var cols = this.getColumns() ;
// one method to search is for each 
// another method is lambda like it is done in validateEntity()  method x.property == Z
cols.forEach((value,index)=> { if(value == "Z" {cols.text = newName}    });
}
```
You can search Northwind project and see examples of each of those methods.. 
Checkbox grid shows column searching and iteration.
validateEntity shows how to lambda search - lambda is nice one liner and easy enough to read and understand ..

search northwind for getColumns() and validateEntity() that should show you enough to do what you need to do.

**Comment by MagDev175** [comment, marked as answer]
Thanks. I actually solved this by using the setColumns method in slick grid. My issue was that getColumns was only called before load and the values are passed from the dialog to the grid after load. 
The below code worked for me:

```
        set reporttype(value: string) {
            if (value.indexOf(" X Report") > 0) {
                var cols = this.getColumns();
                Q.first(cols, x => x.field === fld.Field1).name = "Category"
                this.slickGrid.setColumns(cols);
                
            }
}
```

**Reply by stixoffire** [reply]
Glad you got your issue resolved.
Just as an fyi there are a couple methods I resort to when things do not seem to be working (basically trying to determine when methods are called) .
I use afterEntityLoad() or markupReady()  the last one markupReady()  is great when it comes to passing values into forms and also changing formatting. Helps to isolate issues as a issue of timing or issue of code .. 

also on your setColumns(stuff); method - you might need to re-render your grid to preserve formatting.
grid.render(); that method has helped me resolve some other issues .. that seemed not related but in the end resolved them.
